File: 0092-reverse-linked-list-ii/0092-reverse-linked-list-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int left, int right) {
        if(!head || left == right) return head; 
        ListNode* dummy = new ListNode(-1); 
        dummy->next = head; 
        ListNode* prev = dummy; 
        for(int i = 1; i < left; i++){
            prev = prev->next; 
        }
        ListNode* curr = prev->next; 
        ListNode* nextHeader = nullptr; 

        for(int i = left; i <= right; i++){
            ListNode* tmp = curr->next; 
            curr->next = nextHeader; 
            nextHeader = curr; 
            curr = tmp; 
        }

        prev->next->next = curr; 
        prev->next = nextHeader; 

        return dummy->next; 
    }
};

File: 0092-reverse-linked-list-ii/0092-reverse-linked-list-ii.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseBetween(ListNode head, int left, int right) {
        ListNode dummy = new ListNode(-1); 
        dummy.next = head; 
        ListNode curr = dummy; 
        for(int i = 0; i < left - 1; i++){
            curr = curr.next; 
        }
        
        ListNode tail = curr.next; 
        ListNode newHead = curr.next; 
        ListNode prev = null; 
        for(int i = left; i <= right; i++){
            ListNode tmp = newHead.next; 
            newHead.next = prev; 
            prev = newHead; 
            newHead = tmp; 
        }

        curr.next = prev; 
        tail.next = newHead; 

        return dummy.next; 
    }
}